tests/php/test-class-amp-options-rest-controller.php

```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * Tests for AMP_Options_REST_Controller.
+ *
+ * @package AMP
+ */
+
+use AmpProject\AmpWP\Tests\ThemesApiRequestMocking;
+
+/**
+ * Tests for AMP_Options_REST_Controller.
+ *
+ * @group amp-options
+ *
+ * @covers AMP_Options_REST_Controller
+ */
+class Test_AMP_Options_REST_Controller extends WP_UnitTestCase {
+
+	use ThemesApiRequestMocking;
+
+	/**
+	 * Test instance.
+	 *
+	 * @var AMP_Options_REST_Controller
+	 */
+	private $controller;
+
+	/**
+	 * Set up.
+	 */
+	public function setUp() {
+		parent::setUp();
+
+		$this->add_reader_themes_request_filter();
+
+		do_action( 'rest_api_init' );
+		$this->controller = new AMP_Options_REST_Controller( new AMP_Reader_Themes() );
+	}
+
+	/**
+	 * Tests AMP_Options_REST_Controller::register_routes.
+	 *
+	 * @covers AMP_Options_REST_Controller::register_routes
+	 */
+	public function test_register_routes() {
+		$this->controller->register_routes();
+
+		$this->assertContains( 'amp/v1', rest_get_server()->get_namespaces() );
+		$this->assertContains( '/amp/v1/options', array_keys( rest_get_server()->get_routes( 'amp/v1' ) ) );
+	}
+
+	/**
+	 * Tests AMP_Options_REST_Controller::get_items_permissions_check.
+	 *
+	 * @covers AMP_Options_REST_Controller::get_items_permissions_check
+	 */
+	public function test_get_items_permissions_check() {
+		$this->assertWPError( $this->controller->get_items_permissions_check( new WP_REST_Request( 'GET', '/amp/v1/options' ) ) );
+
+		wp_set_current_user( 1 );
+
+		$this->assertTrue( $this->controller->get_items_permissions_check( new WP_REST_Request( 'GET', '/amp/v1/options' ) ) );
+	}
+
+	/**
+	 * Tests AMP_Options_REST_Controller::get_items.
+	 *
+	 * @covers AMP_Options_REST_Controller::get_items.
+	 */
+	public function test_get_items() {
+		$this->assertEquals(
+			array_keys( $this->controller->get_items( new WP_REST_Request( 'GET', '/amp/v1/options' ) )->get_data() ),
+			[
+				'theme_support',
+				'reader_theme',
+			]
+		);
+	}
+
+	/**
+	 * Tests AMP_Options_REST_Controller::update_items.
+	 *
+	 * @covers AMP_Options_REST_Controller::update_items.
+	 */
+	public function test_update_items() {
+		wp_set_current_user( 1 );
+
+		$request = new WP_REST_Request( 'POST', '/amp/v1/options' );
+		$request->set_body_params(
+			[
+				'reader_theme'  => 'twentysixteen',
+				'theme_support' => 'transitional',
+			]
+		);
+		$response = rest_get_server()->dispatch( $request );
+
+		if ( isset( $response->get_data()['code'] ) ) { // Result is different pre-WP 5.0.
+			$this->assertEquals( 'rest_invalid_param', $response->get_data()['code'] );
```

Comment:
Does this mean that the REST API endpoint should not be even made available in WP<5.0? What is the result here? We intend to not make the settings screen or wizard available in WP 4.9, correct?

Reply:
Currently I don't see a reason for the REST endpoints to be made available pre-5.0. It might be a good idea to wrap them in the same `amp_should_use_new_onboarding` check that prevents the setup wizard from running. I hadn't thought about that for the REST endpoints, but if we don't need them pre-5.0, might as well turn them off.